daemon/network.go

```diff
@@ -99,15 +100,45 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+)
+
+func (daemon *Daemon) setupIngressWorker() {
+	ingressWorkerStop = make(chan struct{}, 1)
+	ingressJobsChannel = make(chan *ingressJob, 100)
+	go func() {
+		for {
+			select {
+			case <-ingressWorkerStop:
+				ingressID = ""
+				return
+			case r := <-ingressJobsChannel:
+				if r.create != nil {
+					daemon.setupIngress(r.create, r.ip, ingressID)
+					ingressID = r.create.ID
+				} else {
+					daemon.releaseIngress(ingressID)
+					ingressID = ""
+				}
+			}
+		}
+	}()
+}
 
-func ingressWait() func() {
-	ingressChan <- struct{}{}
-	return func() { <-ingressChan }
+func (daemon *Daemon) stopIngressWorker() {
+	if ingressWorkerStop != nil {
```

Comment:
> It implies that stopIngressWorker can be called before setupIngressWorker

I see your point. 

The fact is the `SetupIngressNetwork` being called from the executor and the call to `DaemonLeavesCluster` happen from two different threads. Their execution order is not guaranteed, and I saw it not deterministic during the testing of initial diffs.

In fact in the setup ingress call, we need to wait for `agentInitWait()` for this reason.

So it is not excluded a `swarm join` quickly followed by a `swarm leave` would flip the order of events.